modules/videoio/src/cap_android_mediandk.cpp

```diff
@@ -0,0 +1,247 @@
+                        sawOutputEOS = true;
+                    }
+                    AMediaCodec_releaseOutputBuffer(mediaCodec, bufferIndex, info.size != 0);
+                    return true;
+                } else if (bufferIndex == AMEDIACODEC_INFO_OUTPUT_BUFFERS_CHANGED) {
+                    LOGV("output buffers changed");
+                } else if (bufferIndex == AMEDIACODEC_INFO_OUTPUT_FORMAT_CHANGED) {
+                    auto format = AMediaCodec_getOutputFormat(mediaCodec);
+                    LOGV("format changed to: %s", AMediaFormat_toString(format));
+                    AMediaFormat_delete(format);
+                } else if (bufferIndex == AMEDIACODEC_INFO_TRY_AGAIN_LATER) {
+                    LOGV("no output buffer right now");
+                } else {
+                    LOGV("unexpected info code: %zd", bufferIndex);
+                }
+            }
+        }
+        return false;
+    }
+
+    bool isOpened() const CV_OVERRIDE { return mediaCodec != NULL; }
+
+    int getCaptureDomain() CV_OVERRIDE { return CAP_ANDROID; }
+
+    bool grabFrame() CV_OVERRIDE
+    {
+        // clear the previous frame
+        buffer = NULL;
+        bufferSize = 0;
+        return decodeFrame();
+    }
+
+    bool retrieveFrame(int, OutputArray out) CV_OVERRIDE
+    {
+        if (buffer == NULL) {
+            return false;
+        }
+        Mat yuv(frameHeight + frameHeight/2, frameWidth, CV_8UC1, buffer);
+        if (colorFormat == COLOR_FormatYUV420Planar) {
+            cv::cvtColor(yuv, out, cv::COLOR_YUV2BGR_YV12);
+        } else if (colorFormat == COLOR_FormatYUV420SemiPlanar) {
+            cv::cvtColor(yuv, out, cv::COLOR_YUV2BGR_NV21);
+        } else {
+            LOGE("Unsupported video format: %d", colorFormat);
+            return false;
+        }
+        return true;
+    }
+
+    double getProperty(int property_id) const CV_OVERRIDE
+    {
+        switch (property_id)
+        {
+            case CV_CAP_PROP_FRAME_WIDTH: return frameWidth;
+            case CV_CAP_PROP_FRAME_HEIGHT: return frameHeight;
+        }
+        return 0;
+    }
+
+    bool setProperty(int /* property_id */, double /* value */) CV_OVERRIDE
+    {
+        return false;
+    }
+
+    bool initCapture(const char * filename)
+    {
+        struct stat statBuffer;
+        if (stat(filename, &statBuffer) != 0) {
+            LOGE("failed to stat file: %s (%s)", filename, strerror(errno));
+            return false;
+        }
+
+        int fd = open(filename, O_RDONLY);
+
+        if (fd < 0) {
+            LOGE("failed to open file: %s %d (%s)", filename, fd, strerror(errno));
+            return false;
+        }
+
+        AMediaExtractor *extractor = AMediaExtractor_new();
+        media_status_t err = AMediaExtractor_setDataSourceFd(extractor, fd, 0, statBuffer.st_size);
+        close(fd);
+        if (err != AMEDIA_OK) {
+            LOGV("setDataSource error: %d", err);
+            return false;
```

Comment:
Avoid resources lost:
```
AMediaExtractor_delete(extractor);
return false;
```

Consider using of `std::shared_ptr` with custom deleter to avoid such manual resource management, like [here](https://android.googlesource.com/platform/cts/+/ab2f2cb79509485ea2ce1c67dece59daf3bfaa74/tests/tests/media/libmediandkjni/native_media_decoder_source.cpp#86).